x-pack/plugins/security_solution/public/detection_engine/rule_management/components/rule_details/rule_diff_tab.tsx

```diff
@@ -16,22 +16,125 @@ import {
 import { DiffView } from './json_diff/diff_view';
 import * as i18n from './json_diff/translations';
+import { getHumanizedDuration } from '../../../../detections/pages/detection_engine/rules/helpers';
+
+/* Inclding these properties in diff display might be confusing to users. */
+const HIDDEN_PROPERTIES = [
+  /*
+    By default, prebuilt rules don't have any actions or exception lists. So if a user has defined actions or exception lists for a rule, it'll show up as diff. This looks confusing as the user might think that their actions and exceptions lists will get removed after the upgrade, which is not the case - they will be preserved.
+  */
+  'actions',
+  'exceptions_list',
+
+  /*
+    Most prebuilt rules are installed as "disabled". Once user enables a rule, it'll show as diff, which doesn't add value.
+  */
+  'enabled',
+
+  /* Technical property. Hidden because it can be confused with "version" by users. */
+  'revision',
+
+  /*
+    "updated_at" value is regenerated on every '/upgrade/_review' endpoint run 
+    and will therefore always show a diff. It adds no value to display it to the user.
+  */
+  'updated_at',
+];
 
 const sortAndStringifyJson = (jsObject: Record<string, unknown>): string =>
   stringify(jsObject, { space: 2 });
 
+/**
+ * Normalizes the rule object, making it suitable for comparison with another normalized rule.
+ *
+ * Normalization is needed to avoid showing diffs for semantically equal property values.
+ * Saving a rule via the rule editing UI might change the format of some properties or set default values.
+ * This function compensates for those changes.
+ *
+ * @param {RuleResponse} originalRule - Rule of a RuleResponse type.
+ * @returns {RuleResponse} - The updated normalized rule object.
+ */
+const normalizeRule = (originalRule: RuleResponse): RuleResponse => {
+  const rule = { ...originalRule };
+
+  /*
+    Convert the "from" property value to a humanized duration string, like 'now-1m' or 'now-2h'. 
+    Conversion is needed to skip showing the diff for the "from" property when the same 
+    duration is represented in different time units. For instance, 'now-1h' and 'now-3600s' 
+    indicate a one-hour duration.
+    The same helper is used in the rule editing UI to format "from" before submitting the edits. 
+    So, after the rule is saved, the "from" property unit/value might differ from what's in the package.
+  */
+  rule.from = formatScheduleStepData({
+    interval: rule.interval,
+    from: getHumanizedDuration(rule.from, rule.interval),
+    to: rule.to,
+  }).from;
+
+  /*
+    Default "note" to an empty string if it's not present.
+    Sometimes, in a new version of a rule, the "note" value equals an empty string, while 
+    in the old version, it wasn't specified at all (undefined becomes ''). In this case, 
+    it doesn't make sense to show diff, so we default falsy values to ''.
+  */
+  rule.note = rule.note ?? '';
```

Comment:
Have you found an example where the prebuilt rule asset comes with `note` being an empty string? I checked the detection-rules repo and didn't find one, which makes me think it might be a bug on our side.

Anyways, I think that we can prevent this issue from happening by setting a default to `note` on the `convertPrebuiltRuleAssetToRuleResponse` util, so that if the `prebuiltRuleAsset` has `note` as undefined, then it will still be installed as an empty string, and we won't see any diffs with a target rule.

```ts
export const convertPrebuiltRuleAssetToRuleResponse = (
  prebuiltRuleAsset: PrebuiltRuleAsset
): RuleResponse => {
  const prebuiltRuleAssetDefaults = {
    enabled: false,
    risk_score_mapping: [],
    severity_mapping: [],
    interval: DEFAULT_INTERVAL,
    to: DEFAULT_TO,
    from: DEFAULT_FROM,
    exceptions_list: [],
    false_positives: [],
    max_signals: DEFAULT_MAX_SIGNALS,
    actions: [],
    related_integrations: [],
    required_fields: [],
    setup: '',
    note: '', // <========= add this default
    references: [],
    threat: [],
    tags: [],
    author: [],
    
  };

```

Reply:
A couple examples of prebuilt rules with `note` equal empty string: [1](https://github.com/elastic/integrations/blob/4029af97a8c345176860e5df0a6b64bcf4df2731/packages/security_detection_engine/kibana/security_rule/7ceb2216-47dd-4e64-9433-cddc99727623_104.json#L17), [2](https://github.com/elastic/integrations/blob/4029af97a8c345176860e5df0a6b64bcf4df2731/packages/security_detection_engine/kibana/security_rule/14de811c-d60f-11ec-9fd7-f661ea17fbce_203.json#L16). There are many, actually. Also many rules where `note` is not defined at all, like [this one](https://github.com/elastic/integrations/blob/main/packages/security_detection_engine/kibana/security_rule/835c0622-114e-40b5-a346-f843ea5d01f1_4.json#L13).

I think adding defaults during installation is a good idea – it fits nicely with the proposed way to do migrations in Phase 3.

However, one issue I see with this approach is that for rules that were installed before the change, we'll show a diff in these two cases:
 - "note" was undefined and became "" in the update
 - "note" was undefined and stayed undefined in the update (but `convertPrebuiltRuleAssetToRuleResponse` will convert it into "")
To me, it makes sense to skip showing a diff for these cases as there's no real change for a user.

To avoid this, we'll need to also normalize the installed rule - either in UI code (current approach) or on read.
If we go the normalize-on-read way, we could do it in [calculateRuleInfos](https://github.com/nikitaindik/kibana/blob/d7a723a53eca2321cd8a580a9a3d58ef441828fd/x-pack/plugins/security_solution/server/lib/detection_engine/prebuilt_rules/api/review_rule_upgrade/review_rule_upgrade_route.ts#L114).
```typescript
const targetRule: RuleResponse = {
  ...convertPrebuiltRuleAssetToRuleResponse(targetVersion),
  id: installedCurrentVersion.id,
  revision: installedCurrentVersion.revision + 1,
  created_at: installedCurrentVersion.created_at,
  /* ... */
};

return {
 /* ... */
 /* Normalize on read here, instead of in UI */ 
  current_rule: normalizeRule(installedCurrentVersion), // <--
  target_rule: targetRule,
  /* ... */
};
```
Perhaps we might as well move transformations for other properties to the backend. It seems to be a better approach because UI won't have to know about these conversions.

But I'm not sure if it's worth the effort now as we are not planning to use whole JSON rule objects starting from Stage 2. And we are preparing significant changes in this area for Stage 3. It may make sense to leave this logic in UI because it's specific to Stage 1 diffs and can be easily removed as we release Stage 2. wdyt?

Reply:
I just meant: let's do both approaches.

Let's have `rule.note = rule.note ?? '';` in the frontend, to solve the issue in the two cases that you mentioned, but also add the empty string as default in the backend to progressively get rid of this problem in the future.